scrapy/core/spidermw.py

```diff
@@ -60,11 +60,22 @@ def process_spider_exception(_failure):
             return _failure
 
         def process_spider_output(result):
+            def wrapper(result_iterable):
+                try:
+                    for r in result_iterable:
+                        yield r
+                except Exception as ex:
+                    exception_result = process_spider_exception(Failure(ex))
+                    if exception_result is None or isinstance(exception_result, Failure):
+                        raise
+                    for output in exception_result:
+                        yield output
             for method in self.methods['process_spider_output']:
                 result = method(response=response, result=result, spider=spider)
-                assert _isiterable(result), \
-                    'Middleware %s must returns an iterable object, got %s ' % \
-                    (fname(method), type(result))
+                if _isiterable(result):
```

Comment:
[This line](https://github.com/scrapy/scrapy/pull/2061/files#diff-7a382cc9280309e8fc166e4b479deb33R58) checks for an `AssertionError` explicitly raised by any `process_spider_input` method, and prevents it from being passed to the exception handler (the only problem would be an external exception raised with the exact same message the spider middleware manager specifies, I believe that case could be ignored).

There are [some tests here](https://github.com/scrapy/scrapy/pull/2061/files#diff-442dbe445eb9f6625bfb8186556983f2R175) to make sure that invalid return values from either `process_spider_input` or `process_spider_output` do not get handled by `process_spider_exception`.